pennylane/labs/trotter_error/product_formulas/error.py

```diff
@@ -289,3 +480,163 @@ def _group_sums_in_dict(term_dict: dict[tuple[Hashable], complex]) -> list[tuple
+            _AdditiveIdentity(),
+        )
+        return weighted_fragment.norm({"gridpoints": gridpoints}) if weighted_fragment else 0.0
+
+    return fragments[element].norm({"gridpoints": gridpoints}) if element in fragments else 1.0
+
+
+def _setup_probability_distribution(
+    commutators: List[Tuple[Hashable | Set]],
+    fragments: Dict[Hashable, Fragment],
+    timestep: float,
+    gridpoints: int = DEFAULT_GRIDPOINTS,
+) -> np.ndarray:
+    """Setup normalized probability distribution for importance sampling.
+
+    This function calculates and normalizes probabilities for all commutators,
+    providing a foundation for importance sampling methods in future PRs.
+
+    Args:
+        commutators: List of commutator tuples
+        fragments: Fragment dictionary
+        timestep: Time step for probability calculation
+        gridpoints: Number of gridpoints for norm calculations
+
+    Returns:
+        Normalized probability array
+    """
+    # Calculate raw probabilities using vectorized operations
+    probabilities = np.array(
+        [
+            _calculate_commutator_probability(comm, fragments, timestep, gridpoints)
+            for comm in commutators
+        ]
+    )
+
+    # Normalize with fallback to uniform distribution
+    total_prob = np.sum(probabilities)
+    return (
+        probabilities / total_prob
+        if total_prob > 0
+        else np.ones(len(commutators)) / len(commutators)
+    )
+
+
+def _apply_sampling_strategy(
+    commutators: List[Tuple[Hashable | Set]],
+    config: SamplingConfig,
+    probabilities: Optional[np.ndarray] = None,  # pylint: disable=unused-argument
+) -> Tuple[List[Tuple[Hashable | Set]], List[float]]:
+    """Apply sampling strategy to select commutators and compute weights.
+
+    This function provides a unified interface for different sampling methods
+    that will be implemented in future PRs. Currently supports only 'exact' method.
+
+    Args:
+        commutators: List of all available commutators
+        config: Sampling configuration
+        probabilities: Pre-computed probabilities (for importance/top_k methods)
+
+    Returns:
+        Tuple of (selected_commutators, weights)
+
+    Raises:
+        NotImplementedError: For sampling methods not yet implemented
+    """
+    if config.method == "exact":
+        # Return all commutators with uniform weights
+        return commutators, [1.0] * len(commutators)
+
+    # Future PRs will implement these methods
+    if config.method == "random":
+        raise NotImplementedError("Random sampling will be implemented in future PR")
+    if config.method == "importance":
+        raise NotImplementedError("Importance sampling will be implemented in future PR")
+    if config.method == "top_k":
+        raise NotImplementedError("Top-k sampling will be implemented in future PR")
+
+    raise ValueError(f"Unknown sampling method: {config.method}")
+
+
+def _compute_expectation_values_with_cache(
```

Comment:
Function to compute all expectation values using `_get_expval_state` that leverages caching. Perhaps there is a better engineering.